examples-gallery/plot_crane_moving_a_load.py

```diff
@@ -0,0 +1,294 @@
+specified_symbols = (F,)
+h_opty = sm.symbols('h_opty')
+unknown_symbols = []
+methode = "backward euler"
+
+num_nodes = 250
+duration =(num_nodes - 1) * h_opty
+interval_value = h_opty
+
+# Specify the known system parameters.
+par_map = OrderedDict()
+par_map[l] = 5.0
+par_map[m1] = 1.0
+par_map[m2] = 10.0
+par_map[g] = 9.81
+
+t0, tf = 0.0, duration
+
+def obj(free):
+    """Minimize h_opty, the time interval between nodes."""
+    return free[-1]
+
+def obj_grad(free):
+    grad = np.zeros_like(free)
+    grad[-1] = 1.
+    return grad
+
+# start location and and final location of the load.
+anfang = 0.0
+ende = 15.0
+
+# %%
+# Set the initial and final states to form the instance constraints.
+initial_state_constraints = {xc: anfang, xl: anfang, yl: -par_map[l], q: 0.0,
+    uxc: 0.0, uxl: 0.0, uyl: 0.0, u: 0.0}
+final_state_constraints = {xc: ende, xl: ende, yl: -par_map[l], q: 0.0,
+    uxc: 0.0, uxl: 0.0, uyl: 0.0, u: 0.0}
+instance_constraints = (
+) + tuple(
+    xi.subs({t: t0}) - xi_val for xi, xi_val in
+        initial_state_constraints.items()
+) + tuple(
+    xi.subs({t: tf}) - xi_val for xi, xi_val in
+        final_state_constraints.items()
+)
+# %%
+# Forcing h_opty > 0.0 sometimes avoids negative 'solutions'. Here it also
+# seem to help with the convergence of the optimization: Different bounds,
+# e,g, h_opty = (1.e-4, 1.) gives unreasonable results.
+
+bounds = {
+    F: (-20., 20.),
+    xl: (anfang, ende),
+    xc: (anfang, ende),
+    h_opty: (1.e-5, 0.5),
+    }
+
+# Create the optimization problem.
+prob = Problem(
+    obj,
+    obj_grad,
+    EOM,
+    state_symbols,
+    num_nodes,
+    interval_value,
+    known_parameter_map=par_map,
+    instance_constraints=instance_constraints,
+    bounds=bounds,
+    integration_method=methode,
+    )
+# %%
+# Using a 'reasonable' initial guess speeds up the optimization process a lot.
+# 'Bad' initial guesses may even prevent convergence.
+i1 = [(ende - anfang) / num_nodes * i for i in range(num_nodes)]
+i2 = [0. for _ in range(num_nodes)]
+i3 = i1
+i4 = [-par_map[l] for _ in range(num_nodes)]
+i5 = [0. for _ in range(5*num_nodes)]
+i6 = [0.01]
+initial_guess = np.array(i1 + i2 + i3 + i4 + i5 + i6)
+
+# %%
+# Plot the initial guess.
+prob.plot_trajectories(initial_guess)
+
+
+# this allows to change the maximum number of iterations.
+# Standard is 3000.
+prob.add_option('max_iter', 1000)
+# Find the optimal solution.
+for _ in range(1):
+    solution, info = prob.solve(initial_guess)
+    print('message from optimizer:', info['status_msg'])
+    prob.plot_objective_value()
+    initial_guess = solution
+print('Iterations needed', len(prob.obj_value))
+print(f"Objective value {solution[-1]: .3e}")
+
+# %%
+# plot the accuracy of the solution.
+prob.plot_constraint_violations(solution)
+
+# %%
+# Plot the state trajectories.
+fig, ax = plt.subplots(9, 1, figsize=(7.25, 3.0*9), sharex=True)
+prob.plot_trajectories(solution, ax)
+
+# %%
+# Animate the simulation.
+fps = 30
```

Comment:
This is a function of the optimal `h` value. If you choose 30 frames per second, the animation time will not align with the actual solution.

Reply:
> This is a function of the optimal `h` value. If you choose 30 frames per second, the animation time will not align with the actual solution.

I  just copied this idea from Timo. I run the animation for the duration determined by h, that is duration = (num_nodes - 1) * h.
So, I am not sure I understand what you mean.

Reply:
If h = 0.135, for example, then there is 1 frame every 0.135 seconds. If you set the fps to 30 in the animation, there is one frame per 0.05 seconds.

Reply:
> If h = 0.135, for example, then there is 1 frame every 0.135 seconds. If you set the fps to 30 in the animation, there is one frame per 0.05 seconds.

Do you mean, the animation will run in slow motion or in fast motion compared to 'reality'?

NB: I will make the corrections you suggested this evening an push a new version. I will also make these corrections on my other simulations, so don't waste your time looking at them before I pshed new versions.

Reply:
Setting fps to match the spacing in a time vector will cause the animation to at a realistic rate.

Reply:
> Setting fps to match the spacing in a time vector will cause the animation to at a realistic rate.

So, fps approx. = 1 / h, where h = duration / (num_nodes - 1) ?